### PR TITLE
RTL8139 Rx path

### PR DESCRIPTION
**Pull request by komik0**
this is moved and improved #900

additionally: added netdump - userspace program to dump received packets (packet by packet)

this pr depends on #1003 and #1004

### CODE REVIEW

File: bin/netdump/netdump.c [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+int
+main(int argc, char *argv[])
+{
+  FILE *fp = fopen("/dev/rtl8139", "r+b");
+
+  setprogname(argv[0]);
+  (void)setlocale(LC_ALL, "");
+
+  int n = fread(packet, sizeof(uint8_t), PACKET_SIZE, fp);
```
Comment (pj1031999) [inline bin/netdump/netdump.c:27]: Return value for `fread` is `size_t`.

File: sys/drv/rtl8139.c [hunk 2]
```diff
@@ -39,28 +48,79 @@ static int rtl_reset(rtl8139_state_t *state) {
...
+      return 0;
+    }
+    int res = ringbuf_read(&state->dev_rx_buf, uio);
+    /* fix uio values */
+    if (!res)
+      uio->uio_resid = uio_resid - pkt.len;
+    else
+      uio->uio_resid = uio_resid;
+    return res;
+  } else
```
Comment (pj1031999) [inline sys/drv/rtl8139.c:107]: You can remove that `else`.

File: sys/drv/rtl8139.c [hunk 3]
```diff
@@ -39,28 +48,79 @@ static int rtl_reset(rtl8139_state_t *state) {
...
 
+static int rtl8139_read(vnode_t *v, uio_t *uio, int ioflag) {
+  rtl8139_state_t *state = devfs_node_data(v);
+  uio->uio_offset = 0;
+
+  /* TODO: WITH_SPIN/MUTEX_LOCK ? */
+  if (!ringbuf_empty(&state->dev_rx_buf)) {
+    size_t uio_resid = uio->uio_resid;
+    rtl8139_packet_t pkt;
+    ringbuf_getnb(&state->dev_rx_buf, (uint8_t *)&pkt, RL_PKTHDR_SIZE);
```
Comment (pj1031999) [inline sys/drv/rtl8139.c:89]: What if inside `ringbuf_getnb` you get interrupt that triggers `do_receive`? Is it safe to do that without any lock?

**Comment by cahirwpz** [comment]
Since we're going to drop support for Malta in the near future this driver has little to no value for us. Hence I decided to close the PR.

**Pull request closed without merging**